### PR TITLE
Support Windows as remote node.

### PR DESCRIPTION
**Pull request by squirrelsc**
1. Add WindowsCommand.ts to support different shell commands in windows.
1. Update documents to explain how to enable remote node in Windows.

Fixed some bugs,

1. Fix folder support problem on tarAdd function in training_service/common/util.ts.
1. Fix Keras compatibility issue on val_acc or val_accuracy name in logs.
1. Fix killChildProcesses of linux to support kill all descendant processes, instead of only direct child processes.
1. Fix too muchGPU stats commands 

Small improvements

1. Let nni_trial_tool.trial_keeper to save its process id itself, to simplify script, and be more reliable.
1. Migrate non-cross platform code into osCommands.ts.
1. Refine ExecutorManager and ShellExecutor, to let ShellExecutor manage MaxTrials, and make it `1` to reduce concurrency issues.
1. Move map of Trial to shellexecutor from RemoteMachineTrainingService to ExecutorManager. It let resource management be more convenient, and avoid null reference.
1. Add RandomInt in utils.ts.
1. Move gpu collector related files into experiment folder for multi-run.
1. Let OsCommands can normalize folder like `/` to `\` in windows, and versus in Linux.
1. Extend timeout of setMetaddata to 30 seconds for high latency remote machines like in cloud.
1. Fix partial success of pipelines-it-remote-windows.yml
1. Remove gpuNum: 0 in test cases to use GPU.
1. Improved some logs.

### CODE REVIEW

File: src/nni_manager/training_service/remote_machine/remoteMachineData.ts [hunk 1]
```diff
@@ -85,78 +85,85 @@ export class RemoteMachineTrialJobDetail implements TrialJobDetail {
...
+            throw new Error("executor shouldn't be undefined before set!");
+        }
+        this.executorMap.set(id, executor);
 
-    /**
-     * first executor instance is used for gpu collector and host job
-     */
-    public getFirstExecutor(): ShellExecutor {
-        return this.executorArray[0];
+        // init a new executor if no free one.
```
Comment (chicm-ms) [inline src/nni_manager/training_service/remote_machine/remoteMachineData.ts:126]: This comments does not match the code

File: src/nni_manager/training_service/remote_machine/remoteMachineData.ts [hunk 2]
```diff
@@ -85,78 +85,85 @@ export class RemoteMachineTrialJobDetail implements TrialJobDetail {
...
+        if (executor === undefined) {
+            throw new Error("executor shouldn't be undefined before return!");
```
Comment (chicm-ms) [inline src/nni_manager/training_service/remote_machine/remoteMachineData.ts:128]: this block is duplicated with above

File: src/nni_manager/training_service/remote_machine/shellExecutor.ts [hunk 3]
```diff
@@ -73,43 +108,120 @@ class ShellExecutor {
...
+        }
+        return canBeReleased;
+    }
+
     public get getUsedConnectionNumber(): number {
-        return this.usedConnectionNumber;
+        return this.usedCount;
     }
 
     public addUsedConnectionNumber(): void {
```
Comment (chicm-ms) [inline src/nni_manager/training_service/remote_machine/shellExecutor.ts:135]: should this function `addUsedConnectionNumber`  be removed?

File: src/nni_manager/training_service/remote_machine/remoteMachineData.ts (hunk 1 above)
Comment (squirrelsc) [reply src/nni_manager/training_service/remote_machine/remoteMachineData.ts]: fixed

File: src/nni_manager/training_service/remote_machine/remoteMachineData.ts (hunk 2 above)
Comment (squirrelsc) [reply src/nni_manager/training_service/remote_machine/remoteMachineData.ts]: fixed

File: src/nni_manager/training_service/remote_machine/shellExecutor.ts (hunk 3 above)
Comment (squirrelsc) [reply src/nni_manager/training_service/remote_machine/shellExecutor.ts]: fixed

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts [hunk 4]
```diff
@@ -520,43 +526,37 @@ class RemoteMachineTrainingService implements TrainingService {
...
         if (trialJobDetail === undefined) {
             throw new Error(`Can not get trial job detail for job: ${trialJobId}`);
         }
 
         const trialLocalTempFolder: string = path.join(this.expRootDir, 'trials-local', trialJobId);
 
-        await executor.createFolder(trialWorkingFolder);
-        await executor.createFolder(unixPathJoin(trialWorkingFolder, '.nni'));
+        await executor.createFolder(trialJobDetail.workingDirectory);
+        await executor.createFolder(executor.joinPath(trialJobDetail.workingDirectory, '.nni'));
```
Comment (SparkSnail) [inline src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts:544]: Why not use a function support create folder recursively?

File: src/nni_manager/training_service/remote_machine/test/shellExecutor.test.ts [hunk 5]
```diff
@@ -12,25 +12,28 @@ import { Client } from 'ssh2';
...
-        await executor.copyFileToRemote(LOCALFILE, REMOTEFILE);
+        await executor.copyFileToRemote(LOCALFILE, remoteFullName);
     }
 }
 
 async function getRemoteFileContentLoop(executor: ShellExecutor): Promise<void> {
-    for (let i: number = 0; i < 10; i++) {
+    const remoteFullName = executor.joinPath(executor.getTempPath(), REMOTEFILE);
+    for (let i: number = 0; i < 3; i++) {
         // console.log(i);
```
Comment (SparkSnail) [inline src/nni_manager/training_service/remote_machine/test/shellExecutor.test.ts:35]: remove this line

File: src/nni_manager/training_service/remote_machine/test/shellExecutor.test.ts [hunk 6]
```diff
@@ -41,14 +44,16 @@ describe('ShellExecutor test', () => {
         rmMeta = JSON.parse(fs.readFileSync('../../.vscode/rminfo.json', 'utf8'));
         console.log(rmMeta);
     } catch (err) {
-        console.log(`Please configure rminfo.json to enable remote machine test.${err}`);
+        console.log(`Please configure rminfo.json to enable remote machine test. ${err}`);
```
Comment (SparkSnail) [inline src/nni_manager/training_service/remote_machine/test/shellExecutor.test.ts:47]: remove console.log() in TS code, use `this.log()` instead.

File: test/config/examples/cifar10-pytorch.yml [hunk 7]
```diff
@@ -14,7 +14,6 @@ assessor:
 trial:
   codeDir: ../../../examples/trials/cifar10_pytorch
   command: python3 main.py --epochs 1 --batches 1
-  gpuNum: 1
```
Comment (SparkSnail) [inline test/config/examples/cifar10-pytorch.yml:17]: Why remove gpuNum?

File: src/nni_manager/training_service/remote_machine/test/shellExecutor.test.ts (hunk 5 above)
Comment (squirrelsc) [reply src/nni_manager/training_service/remote_machine/test/shellExecutor.test.ts:35]: fixed

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts (hunk 4 above)
Comment (squirrelsc) [reply src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts:543]: fixed

File: src/nni_manager/training_service/remote_machine/test/shellExecutor.test.ts (hunk 6 above)
Comment (squirrelsc) [reply src/nni_manager/training_service/remote_machine/test/shellExecutor.test.ts:47]: tried, it doesn't work in test code.

File: test/config/examples/cifar10-pytorch.yml (hunk 7 above)
Comment (squirrelsc) [reply test/config/examples/cifar10-pytorch.yml:17]: It looks the test environment is used by two test suites sometime. So if set it to 1, NNI will wait another finished, but it causes timeout error. Remove this setting, it lets test cases to use default GPU, not wait each other, and reduced failure chance.

Comment (SparkSnail) [reply test/config/examples/cifar10-pytorch.yml:17]: We need to add examples to test gpuScheduler in NNI, if remove this configuration, we could not test gpu related functions in pipeline.

Comment (squirrelsc) [reply test/config/examples/cifar10-pytorch.yml:17]: ok, I will add it back. But it's test is very limited.
